core/src/main/java/com/linecorp/armeria/server/file/FileService.java

```diff
@@ -344,19 +372,40 @@ private HttpFile findFile(ServiceRequestContext ctx, HttpRequest req) {
 
             // Cache hit, but the cached file is out of date. Replace the old entry from the cache.
             cache.invalidate(pathAndEncoding);
-            return cache(ctx, pathAndEncoding, uncachedFile);
+            return cache(ctx, pathAndEncoding, uncachedFile, encoding, decompress);
         });
     }
 
-    private HttpFile cache(
-            ServiceRequestContext ctx, PathAndEncoding pathAndEncoding, HttpFile uncachedFile) {
+    private CompletableFuture<@Nullable HttpFile> findFileAndDecompress(
+            ServiceRequestContext ctx, String path, Set<ContentEncoding> supportedEncodings) {
+        // Look up a non-compressed file first to avoid extra decompression
+        return findFile(ctx, path, (ContentEncoding) null, false).thenCompose(file -> {
+            if (file != null) {
+                return UnmodifiableFuture.completedFuture(file);
+            } else {
+                return findFile(ctx, path, supportedEncodings.iterator(), true);
+            }
+        });
+    }
+
+    private HttpFile cache(ServiceRequestContext ctx, PathAndEncoding pathAndEncoding, HttpFile uncachedFile,
+                           @Nullable ContentEncoding encoding, boolean decompress) {
 
         assert cache != null;
 
         final Executor executor = ctx.blockingTaskExecutor();
         final ByteBufAllocator alloc = ctx.alloc();
 
         return HttpFile.from(uncachedFile.aggregateWithPooledObjects(executor, alloc).thenApply(aggregated -> {
+            if (decompress && encoding != null) {
+                aggregated = decompress(aggregated, encoding, alloc);
+                if (aggregated.attributes().length() > config.maxCacheEntrySizeBytes()) {
```

Comment:
`aggregated.attributes()` is `null` when `aggregated` is `AggregatedHttpFile.nonExistent()`

Reply:
A non-existent file should not reach here and `decompress()` also checks `attributes`.
https://github.com/line/armeria/blob/c0acac56b8258f05b99f80aa8b11ad3959163ac3/core/src/main/java/com/linecorp/armeria/server/file/FileService.java#L428-L430
For clarification, let me add additional assertions.